app/src/main/java/com/github/odaridavid/weatherapp/data/weather/DefaultWeatherRepository.kt

```diff
@@ -1,57 +1,85 @@
+import kotlinx.coroutines.flow.onCompletion
+import timber.log.Timber
 import java.io.IOException
 import java.net.HttpURLConnection.HTTP_UNAUTHORIZED
+import java.util.concurrent.TimeUnit
 import javax.inject.Inject
 
 class DefaultWeatherRepository @Inject constructor(
     private val openWeatherService: OpenWeatherService,
-    private val logger: Logger
+    private val weatherDao: WeatherDao,
+    @ApplicationContext private val context: Context
 ) : WeatherRepository {
     override fun fetchWeatherData(
         defaultLocation: DefaultLocation,
         language: String,
         units: String
     ): Flow<ApiResult<Weather>> = flow {
-
+        val cachedWeather = weatherDao.getWeather()
+        if (cachedWeather != null && cachedWeather.isDataValid()) {
+            val weatherData = cachedWeather.toCoreEntity(unit = units)
+            emit(ApiResult.Success(data = weatherData))
+            return@flow
+        }
         val excludedData = "${ExcludedData.MINUTELY.value},${ExcludedData.ALERTS.value}"
 
-        val response = openWeatherService.getWeatherData(
+        val apiResponse = openWeatherService.getWeatherData(
             longitude = defaultLocation.longitude,
             latitude = defaultLocation.latitude,
             excludedInfo = excludedData,
             units = units,
             language = getLanguageValue(language),
             appid = BuildConfig.OPEN_WEATHER_API_KEY
         )
+        val response = apiResponse.body()
+        if (apiResponse.isSuccessful && response != null) {
+            val entity = apiResponse.body()!!.toWeatherEntity()
 
-        if (response.isSuccessful && response.body() != null) {
-            val weatherData = response.body()!!.toCoreModel(unit = units)
-            emit(ApiResult.Success(data = weatherData))
+            weatherDao.insertCurrentWeather(entity)
+            val weatherData = entity.toCoreEntity(unit= units)
+            emit(ApiResult.Success(data =weatherData))
         } else {
-            val errorMessage = mapResponseCodeToErrorMessage(response.code())
+            val errorMessage = mapResponseCodeToErrorMessage(apiResponse.code())
+            Timber.e("Error Message $errorMessage")
             emit(ApiResult.Error(errorMessage))
         }
     }.catch { throwable ->
         val errorMessage = when (throwable) {
             is IOException -> R.string.error_connection
             else -> R.string.error_generic
         }
-        logger.logException(throwable)
         emit(ApiResult.Error(errorMessage))
+    }.onCompletion {
+        val constraints = Constraints.Builder()
+            .setRequiredNetworkType(NetworkType.CONNECTED)
+            .build()
+
+        val refreshWeatherRequest = OneTimeWorkRequestBuilder<UpdateWeatherWorker>()
+            .setConstraints(constraints)
+            .setInitialDelay(15, TimeUnit.MINUTES)
+            .build()
+
+        WorkManager.getInstance(context).enqueue(refreshWeatherRequest)
```

Comment:
Nice work here👌, but currently it looks like the repository is doing more than intended, you could make use of things like use cases or come up with an approach that abstracts the work manager setup and use.

Reply:
I can make do the use-cases and make the changes